alphastats/llm/llm_functions.py

```diff
@@ -1,15 +1,116 @@
+    This function searches for a feature in the feature-to-representation map that matches the given feature representation.
+    If a matching feature is found in the annotation store, its corresponding annotation is returned.
+    Args:
+        feature_repr (str): The representation of the feature to search for.
+        feature_to_repr_map (dict): A dictionary mapping features to their representations.
+        annotation_store (dict): A dictionary storing annotations for features.
+    Returns:
+        Union[dict, str, None]: The annotation corresponding to the feature representation if found, otherwise None.
+    """
+    for feature in features:
+        if feature in annotation_store:
+            return annotation_store[feature]
+    return None
+
+
+def get_annotation_from_uniprot_by_feature_list(features: list) -> Union[dict, str]:
+    """
+    Retrieve annotation from UniProt based on a list of features.
+    It retrieves the annotation for the most appropriate feature from UniProt. Most appropriate is defined as 1. the only feature, 2. any feature if they only differ in isoform ids, 3. the feature with the most base ids, as this is most likely to contain a well annotated protein.
+    Args:
+        features (list): A list of features to search for.
+    Returns:
+        Union[dict, str]: The annotation for the identified feature. The return type can be a dictionaryn or a string depending on the annotation retrieved.
+    """
+    if len(features) == 1:
+        feature = features[0]
+    elif len(features) == 0:
+        raise ValueError("No features provided.")
+    else:
+        baseid_sets = []
+        for feature in features:
+            baseid_sets.append(set([el.split("-")[0] for el in feature.split(";")]))
+        if len(set.union(*baseid_sets)) == len(set.intersection(*baseid_sets)):
+            feature = features[0]
+        else:
+            feature = features[baseid_sets.index(max(baseid_sets, key=len))]
+    annotation = get_annotations_for_feature(feature)
+    st.session_state[StateKeys.ANNOTATION_STORE][feature] = annotation
+    return annotation
+
+
+def get_uniprot_info_for_llm_input(llm_input: str) -> str:
+    """Get the UniProt information from llm input. This can be either a feature representation, a gene identifier or a protein id.
+    This is required so the LLM can feed the promt from the list of feature representations it is provided with.
+
+    Args:
+        llm_input (str): Either a feature representation, feature id, gene symbol or protein identifier.
+
+    Returns:
+        str: The formatted UniProt information for the feature."""
+
+    feature_to_repr_map = st.session_state[StateKeys.DATASET]._feature_to_repr_map
+    gene_to_features_map = st.session_state[StateKeys.DATASET]._gene_to_features_map
+    protein_to_features_map = st.session_state[
+        StateKeys.DATASET
+    ]._protein_to_features_map
+    annotation_store = st.session_state[StateKeys.ANNOTATION_STORE]
```

Comment:
If it is possible to pass dictionaries to the LLM tools, yes. Otherwise not. This would need to be defined in get_general_assistant_functions. I couldn't find the documentation for allowed types.

Reply:
I added it to the board.